yb-voyager/cmd/exportData.go

```diff
@@ -788,14 +855,258 @@ func getInitialTableList() (map[string]string, []sqlname.NameTuple) {
+		if err != nil {
+			return nil, nil, fmt.Errorf("error getting name  tuple for the string oject name: %v", err)
+		}
+		firstRunTableWithLeafsAndRoots = append(firstRunTableWithLeafsAndRoots, tuple)
+	}
+
+	if source.DBType == YUGABYTEDB {
+		//For the first run of export data from target we will fetch the leaf partitions from target and store them
+		if len(msr.TargetExportedTableListWithLeafPartitions) == 0 || msr.TargetRenameTablesMap == nil {
+			//Now add the leaf partitions to the stored table-list for the first run of export data from target
+			// as it will only have root table names and get the partitionsToRootTableMap
+			partitionsToRootTableMap, firstRunTableWithLeafsAndRoots, err = addLeafPartitionsInTableList(firstRunTableWithLeafsAndRoots, true)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to add the leaf partitions in table list: %w", err)
+			}
+		}
+	}
+	return firstRunTableWithLeafsAndRoots, partitionsToRootTableMap, nil
+}
+
+func getInitialTableList() (map[string]string, []sqlname.NameTuple, error) {
+	// store table list after filtering unsupported or unnecessary tables
+	msr, err := metaDB.GetMigrationStatusRecord()
+	if err != nil {
+		utils.ErrExit("error fetching migration status record: %v", err)
+	}
+	storedTableListNotAvailable := func() bool {
+		//if any of the list in DB is empty
+		return len(msr.TableListExportedFromSource) == 0 || len(msr.SourceExportedTableListWithLeafPartitions) == 0
 	}
 
-	return partitionsToRootTableMap, finalTableList
+	if bool(startClean) || storedTableListNotAvailable() {
+		// fresh start case or the first run where we don't have a table list stored in msr
+		return fetchTablesNamesFromSourceAndFilterTableList()
+
+	}
+
+	//sunsequent run case where we will use a table-list stored in msr
+	var firstRunTableWithLeafsAndRoots []sqlname.NameTuple
+	var partitionsToRootTableMap map[string]string
+
+	firstRunTableWithLeafsAndRoots, partitionsToRootTableMap, err = retrieveFirstRunListAndPartitionsRootMap(msr)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting the first run table list and partition to root mapping: %v", err)
+	}
+
+	//guardrails around the table-list in case of re-run
+
+	registeredList, err := getRegisteredNameRegList()
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting name registry list: %v", err)
+	}
+
+	rootTables := make([]sqlname.NameTuple, 0)
+	for _, v := range partitionsToRootTableMap {
+		tuple, err := namereg.NameReg.LookupTableName(v)
+		if err != nil {
+			utils.ErrExit("look up failed for the table name: %s: %v", v, err)
+		}
+		rootTables = append(rootTables, tuple)
+	}
+	rootTables = lo.UniqBy(rootTables, func(t sqlname.NameTuple) string {
+		return t.ForKey()
+	})
+
+	// Finding all the partitions of all root tables part of migration, and report if there any new partitions added
+	rootToNewLeafTablesMap, err := detectNewLeafPartitionsOnPartitionedTables(rootTables, registeredList)
```

Comment:
should we move this block below to where we are actually checking `rootToNewLeafTablesMap` and doing guardrails?

Reply:
we use the rootToNewLeafTablesMap to calculate the currentList